### PR TITLE
feat: implement flysystem for S3

### PR DESCRIPTION
**Pull request by mussbach**
### Ticket
DPLAN-3415
or
https://www.dev.diplanung.de/DefaultCollection/EfA%20DiPlanung/_workitems/edit/8179

This PR documents implements Flysystem in dplan as a flexible adapter for file actions. Initially there are two storages: local and s3. You may switch  them by setting `FILES_SOURCE=s3.storage` or `FILES_SOURCE=local.storage` respectively in your .env.local.
To enable local s3 buckets you need to enable minio with `containers.minio.enable: true` and `containers.minio_mc.enable: true` in your `.defaults.yml`.

To check the local s3 files you may use the minio ui at `http://localhost:9005`. You can find the login in your docker-compose.yml. You need to create a bucket called `dplan-files` in the ui

I tried to keep this PR as small as possible, but as every file related action needs to be checked and eventually adjusted it is quite large. I am happy to assist when any questions raise.

### How to review/test
You may first run your local instance with the default `FILES_SOURCE=local.storage` and check that any existing files can still be found and any file related action works. You may then copy your local files to the local s3 bucket with `bin/<project> dplan:files:move local s3 --no-delete` and change `FILES_SOURCE=s3.storage` and test again that any file is found and any file related action works

### PR Checklist

Delete the checkbox if it doesn't apply/isn't necessary.

- [x] Tests updated/created
- [ ] Update documentation
- [x] Link all relevant tickets
- [x] Move the tickets on the board accordingly
- [ ] Data-Cy attributes added/updated ([conventions](https://dplan-documentation.demos-europe.eu/development/guidelines-conventions/coding-styleguides/twig_html.html#guideline-for-naming-cypress-hooks))

### CODE REVIEW

File: demosplan/DemosPlanCoreBundle/Command/WebfolderCreateCommand.php
Comment (sierrazamudiodemos) [inline demosplan/DemosPlanCoreBundle/Command/WebfolderCreateCommand.php:1]: Why was this removed?

**Comment by sierrazamudiodemos** [comment]
How can the default storage being defined per project?

File: demosplan/DemosPlanCoreBundle/Command/WebfolderCreateCommand.php
Comment (mussbach) [reply demosplan/DemosPlanCoreBundle/Command/WebfolderCreateCommand.php:1]: This is not needed any more for quite some time and instead of fixing it I removed the dead code

**Comment by mussbach** [comment]
The Flysystemadapter needs to be used in any case where files are used that are not only temporary or need to be local e.g. because they are needed by php funktions to be onsite. This is why I marked all file usages when they are valid to be local with a comment to clarify, that no flysystem adapter needs to be used here.

4. Was a bug that is fixed now
5. Regarding the questions:
- the default storage is defined in .env and may be adjusted in .env.local
- what kind of caching do you mean? The browser has its own cache and afaik flysystem also uses some internal caches
- see above. These are only markers for future devs that local file usage is valid in those cases

**Comment by mussbach** [comment]
True, you need some (any) Value for this. My .env.local contains following:
S3_REGION=us-east-1
S3_ENDPOINT=http://minio:9000
S3_USE_PATH_STYLE_ENDPOINT=true
S3_KEY=<value from docker-compose>
S3_SECRET=<value from docker-compose>
#may be local.storage or s3.storage
FILES_SOURCE=s3.storage

**Comment by sierrazamudiodemos** [comment]
Hey,
1. When I go to main branch and I upload a file
2. then I change to this branch
3. Then I make the env.local changes and also the default.yml changes. On env.local I keep it like: `FILES_SOURCE=local.storage`
4. Then I go to check if file is there
5. then file is not there anymore

I dont run this command  dplan:files:move local s3 --no-delete because I  thought that if I set `FILES_SOURCE=local.storage` then the files are taken anyway from the local.storage

**Comment by mussbach** [comment]
@sierrazamudiodemos thank you, found and fixed

**Review by sierrazamudiodemos — APPROVED** [review]
Thanks, it works!

File: demosplan/DemosPlanCoreBundle/Controller/Document/DemosPlanDocumentController.php [hunk 1]
```diff
@@ -782,9 +790,10 @@ public function elementAdminImportAction(
         }
 
         $extractDir = $this->getElementImportDir($procedureId, $currentUser->getUser());
-        $fn = $uploadedFileInfo->getAbsolutePath();
+        // zip file needs to be local to extract files. So move it to the import dir
+        $uploadedZipFileLocal = $fileService->ensureLocalFileFromHash($uploadedFileInfo->getHash());
```
[the remark this answers is not in the document]
Comment (mussbach) [reply demosplan/DemosPlanCoreBundle/Controller/Document/DemosPlanDocumentController.php:794]: Definitely! the php zip functions require the zip file to be accessible via path, streams are not allowed to be passed into the functions. This is why we need to ensure that the remote file is downloaded locally. Same holds true for several php functions like image functions or the like

File: demosplan/DemosPlanCoreBundle/Controller/Segment/SegmentController.php [hunk 2]
```diff
@@ -135,7 +136,17 @@ public function importSegmentsFromXlsx(
             $file = $fileService->getFileInfo($uploadHash);
             $fileName = $file->getFileName();
             try {
-                $result = $importer->importFromFile($file);
+                $localPath = $fileService->ensureLocalFile($file->getAbsolutePath());
```
[the remark this answers is not in the document]
Comment (mussbach) [reply demosplan/DemosPlanCoreBundle/Controller/Segment/SegmentController.php:139]: yes, exactly. After the action is done, the temporary local file copy could be deleted again

File: demosplan/DemosPlanCoreBundle/Logic/DemosFilesystem.php [hunk 3]
```diff
@@ -10,74 +10,8 @@
 
 namespace demosplan\DemosPlanCoreBundle\Logic;
 
-use RuntimeException;
 use Symfony\Component\Filesystem\Filesystem;
 
 class DemosFilesystem extends Filesystem
 {
```
[the remark this answers is not in the document]
Comment (mussbach) [reply demosplan/DemosPlanCoreBundle/Logic/DemosFilesystem.php:16]: I wanted to keep this PR as small as possible. All calls of DemosFilesystem could (and should) now be refactored to directly call Filesystem
